substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfInfoSectionImpl.java

```diff
@@ -654,43 +655,47 @@ private int writeField(DebugContext context, StructureTypeEntry entry, FieldEntr
 
     private int writeMethodDeclarations(DebugContext context, ClassEntry classEntry, byte[] buffer, int p) {
         int pos = p;
-        List<PrimaryEntry> classPrimaryEntries = classEntry.getPrimaryEntries();
-        for (PrimaryEntry primaryEntry : classPrimaryEntries) {
-            Range range = primaryEntry.getPrimary();
-            /*
-             * Declare all methods including deopt targets even though they are written in separate
-             * CUs.
-             */
-            pos = writeMethodDeclaration(context, classEntry, range, buffer, pos);
+        for (MethodEntry method : classEntry.getMethods()) {
+            if (method.isInImage()) {
+                /*
+                 * Declare all methods including deopt targets even though they are written in
+                 * separate CUs.
+                 */
+                pos = writeMethodDeclaration(context, classEntry, method, buffer, pos);
+            }
         }
 
         return pos;
     }
 
-    private int writeMethodDeclaration(DebugContext context, ClassEntry classEntry, Range range, byte[] buffer, int p) {
+    private int writeMethodDeclaration(DebugContext context, ClassEntry classEntry, MethodEntry method, byte[] buffer, int p) {
         int pos = p;
-        String methodKey = range.getSymbolName();
+        String methodKey = method.getSymbolName();
         setMethodDeclarationIndex(classEntry, methodKey, pos);
-        int modifiers = range.getModifiers();
+        int modifiers = method.getModifiers();
         boolean isStatic = Modifier.isStatic(modifiers);
         log(context, "  [0x%08x] method declaration %s", pos, methodKey);
         int abbrevCode = (isStatic ? DwarfDebugInfo.DW_ABBREV_CODE_method_declaration2 : DwarfDebugInfo.DW_ABBREV_CODE_method_declaration1);
         log(context, "  [0x%08x] <2> Abbrev Number %d", pos, abbrevCode);
         pos = writeAbbrevCode(abbrevCode, buffer, pos);
         log(context, "  [0x%08x]     external  true", pos);
         pos = writeFlag((byte) 1, buffer, pos);
-        String name = uniqueDebugString(range.getMethodName());
+        String name = uniqueDebugString(method.methodName());
         log(context, "  [0x%08x]     name 0x%x (%s)", pos, debugStringIndex(name), name);
         pos = writeAttrStrp(name, buffer, pos);
         int fileIdx = classEntry.localFilesIdx();
```

Comment:
The next few lines need rebasing against head. I suggest this

        FileEntry fileEntry = method.getFileEntry();
        int fileIdx;
        if (fileEntry != null) {
            fileIdx = classEntry.localFilesIdx(fileEntry);
        } else {
            fileEntry = classEntry.getFileEntry();
            fileIdx = classEntry.localFilesIdx();
        }
        log(context, "  [0x%08x]     file 0x%x (%s)", pos, fileIdx, fileEntry.getFullName());

Reply:
I prefer:
```
        FileEntry fileEntry = method.getFileEntry();
        if (fileEntry == null) {
            fileEntry = classEntry.getFileEntry();
        }
        int fileIdx = classEntry.localFilesIdx(fileEntry);
```
If there isn't anything wrong with it.

Reply:
Sure, it comes to the same thing.